<launch>
    <arg name="bag_file" default="/home/leevi/Documents/3d-panoptic/data/kitti/2011_09_26/2011_09_26_drive_0005_sync.bag"/>

    <arg name="visualise_global_ids" default="true"/>
    <arg name="confidence_threshold" default="0.5"/>
    <arg name="iou_threshold"        default="0.4"/>
    <arg name="dbscan_min_points"    default="10"/>
    <arg name="dbscan_epsilon"       default="1.5"/>

    <node pkg="rosbag"
          type="play"
          name="player"
          output="screen"
          args="--clock --rate 0.05 $(arg bag_file)"
          launch-prefix="bash -c 'sleep 5; $0 $@'"/>

    <node pkg="nodelet" type="nodelet" name="pc_nodelet_manager" args="manager" required="true"/>

    <node pkg="nodelet"
          type="nodelet"
          name="stereo_disparity"
          args="load stereo_image_proc/disparity pc_nodelet_manager"
          required="true">

          <remap from="left/image_rect"   to="/cam00/image_raw"/>
          <remap from="left/camera_info"  to="/cam00/camera_info"/>
          <remap from="right/image_rect"  to="/cam01/image_raw"/>
          <remap from="right/camera_info" to="/cam01/camera_info"/>

          <param name="prefilter_size" value="15" />
          <param name="prefilter_cap"  value="31" />

          <param name="correlation_window_size" value="35" />
          <param name="min_disparity"           value="0" />
          <param name="disparity_range"         value="32" />

          <param name="uniqueness_ratio"  value="0.5" />
          <param name="texture_threshold" value="0.5" />
          <param name="speckle_size"      value="10" />
          <param name="speckle_range"     value="2" />

          <param name="approximate_sync" value="true" />
          <param name="queue_size" value="100" />
    </node>

    <node pkg="nodelet"
          type="nodelet"
          name="stereo_to_cloud"
          args="load stereo_image_proc/point_cloud2 pc_nodelet_manager"
          required="true">

          <remap from="left/image_rect_color"  to="/cam02/image_raw"/>
          <remap from="left/camera_info"       to="/cam02/camera_info"/>
          <remap from="right/image_rect_color" to="/cam03/image_raw"/>
          <remap from="right/camera_info"      to="/cam03/camera_info"/>

          <remap from="points2" to="/xyz_rgb_pointcloud"/>

          <param name="approximate_sync" value="true" />
          <param name="queue_size" value="100" />
    </node>

    <node name="voxblox_node" pkg="voxblox_ros" type="tsdf_server" output="screen" args="-alsologtostderr" clear_params="true" launch-prefix="">
        <!-- Pointcloud ROS topic -->
        <remap from="pointcloud" to="/velodyne_points"/>

        <!-- Mesher parameters -->
        <param name="tsdf_voxel_size"         value="0.2" />
        <param name="truncation_distance"     value="0.5" />
        <param name="color_mode"              value="color" />
        <param name="enable_icp"              value="false" />
        <param name="icp_refine_roll_pitch"   value="false" />
        <param name="update_mesh_every_n_sec" value="1.0" />
        <param name="mesh_min_weight"         value="1e-4" />
        <param name="method"                  value="fast" />
        <param name="max_ray_length_m"        value="15.0" />
        <param name="use_const_weight"        value="true" />
        <param name="verbose"                 value="true" />

        <param name="pointcloud_queue_size" value="0" />
        <param name="min_time_between_msgs_sec" value="1" />
        <!-- Mesh save path -->
        <param name="mesh_filename" value="/home/leevi/Documents/RTAB-Map/results/3d_rgb.ply" />

        <!-- Transforms from map to camera -->
        <param name="use_tf_transforms" value="true" />
        <param name="world_frame"       value="world"/>
        <param name="sensor_frame"      value="velodyne"/>
    </node>

    <arg name="rviz_cfg" default="$(find panoptic-3d)/launch/fusion_from_bag.rviz" />

    <!-- Visualization RVIZ -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rviz_cfg)" launch-prefix="">

    </node>

</launch>
